numpy/_core/src/common/raii_utils.hpp

```diff
@@ -126,9 +126,11 @@ class SaveThreadState
 //
 // Instead of
 //
+//   Py_INCREF(descr);
 //   npy_string_allocator *allocator = NpyString_acquire_allocator(descr);
 //   [code that uses allocator]
 //   NpyString_release_allocator(allocator);
+//   Py_DECREF(descr);
```

Comment:
Hmmm, yeah, maybe this is better. OTOH it seems to me like if the descriptor gets cleaned up before the lock that seems very awkward and likely very wrong.

OTOH... I guess it is plausible to write a function that uses immediate C-cleanup on error of the array (and thus descriptor), before the scope is finished.